lib/statics.ts

```diff
@@ -0,0 +1,197 @@
+import Generator from './mapping'
+import { ApiResponse, RequestBody } from '@elastic/elasticsearch/lib/Transport'
+import { Search } from '@elastic/elasticsearch/api/requestParams'
+
+export async function createMapping(this: MongoosasticModel<MongoosasticDocument>, body: RequestBody): Promise<Record<PropertyName, Property>> {
+
+	const options = this.esOptions()
+	const client = this.esClient()
+	
+	const indexName = getIndexName(this)
+	
+	const generator = new Generator()
+	const completeMapping = generator.generateMapping(this.schema)
+
+	const filtered = filterMappingFromMixed(completeMapping.properties)
+	completeMapping.properties = filtered
+
+	const properties = options.properties
+	if (properties) {
+		Object.keys(properties).map(key => {
+			completeMapping.properties[key] = properties[key]
+		})
+	}
+
+	const exists = await client.indices.exists({
+		index: indexName
+	})
+
+	if (exists.body) {
+		await client.indices.putMapping({
+			index: indexName,
+			body: completeMapping
+		})
+		return completeMapping
+	}
+
+	await client.indices.create({
+		index: indexName,
+		body: body
+	})
+
+	await client.indices.putMapping({
+		index: indexName,
+		body: completeMapping
+	})
+
+	return completeMapping
+
+}
+
+export function synchronize(this: MongoosasticModel<MongoosasticDocument>, query: FilterQuery<MongoosasticDocument> = {}, inOpts: SynchronizeOptions = {}): events {
+
+	const options = this.esOptions()
+
+	const em = new events.EventEmitter()
+	let counter = 0
+
+	// Set indexing to be bulk when synchronizing to make synchronizing faster
+	// Set default values when not present
+	const bulkOptions = options.bulk
+	options.bulk = {
+		delay: (options.bulk && options.bulk.delay) || 1000,
+		size: (options.bulk && options.bulk.size) || 1000,
+		batch: (options.bulk && options.bulk.batch) || 50
+	}
+
+	const saveOnSynchronize = inOpts.saveOnSynchronize !== undefined ? inOpts.saveOnSynchronize : options.saveOnSynchronize
+
+	const stream = this.find(query).batchSize(options.bulk.batch).cursor()
+
+	stream.on('data', doc => {
+		stream.pause()
+		counter++
+
+		function onIndex (indexErr: unknown, inDoc: MongoosasticDocument) {
+			counter--
+			if (indexErr) {
+				em.emit('error', indexErr)
+			} else {
+				em.emit('data', null, inDoc)
+			}
+			stream.resume()
+		}
+
+		doc.on('es-indexed', onIndex)
+		doc.on('es-filtered', onIndex)
+
+		if(saveOnSynchronize){
+			doc.save((err: unknown) => {
+				if (err) {
+					counter--
+					em.emit('error', err)
+					return stream.resume()
+				}
+			})
+		} else {
+			postSave(doc)
+		}
+	})
+
+	stream.on('close', () => {
+		const closeInterval = setInterval(() => {
+			if (counter === 0) {
+				clearInterval(closeInterval)
+				em.emit('close')
+				options.bulk = bulkOptions
+			}
+		}, 1000)
+	})
+
+	stream.on('error', err => {
+		em.emit('error', err)
+	})
+
+	return em
+}
+
+export async function esTruncate(this: MongoosasticModel<MongoosasticDocument>): Promise<void> {
```

Comment:
This approach of deleting records is not correct and bad as it will take infinite time on big index, I have changed this to be deleting the index and recreating it with same mapping.

Reply:
The same function is currently used. I tried not to change the methods behavior for this first PR as the main purpose was immigrating to TS.
Yes, i like this approach of deleting the index and recreating it with the same mapping.

Reply:
> > > Also using existing es client is critical function in my opinion, I think we should add it.
> > 
> > 
> > As i already mentioned, it would be hard to maintain the code in that case, as the different es client versions implement different interfaces, so it may broke the package.
> 
> Yes I understand, but the code doesn't need to handle all clients version, the dependency here is X and if they use the new version of the lib would either can use built in client or pass client with same version to it.

I agree with @meabed here. opening the option to pass existing `esClient` is desirable here, even at the risk that the user pass a non-compatible client that is their responsibility. But as @yahiakr says, we can do this in `develop` branch after merging this first migration.